S/SuiteSparse/common.jl

```diff
@@ -71,17 +71,107 @@ products = [
     LibraryProduct("libspqr",                :libspqr),
 ]
 
-# Products for the GPU builds of SuiteSparse
-gpu_products = [
-    LibraryProduct("libcholmod_cuda",                :libcholmod),
-    LibraryProduct("libspqr_cuda",                   :libspqr),
-    LibraryProduct("libgpuqrengine_cuda",            :libgpuqrengine),
-    LibraryProduct("libsuitesparse_gpuruntime_cuda", :libsuitesparse_gpuruntime),
-]
-
 # Dependencies that must be installed before this package can be built
 dependencies = [
     Dependency("libblastrampoline_jll"; compat="5.8.0"),
     BuildDependency("LLVMCompilerRT_jll",platforms=[Platform("x86_64", "linux"; sanitize="memory")]),
     HostBuildDependency(PackageSpec(; name="CMake_jll", version = v"3.24.3"))
 ]
+
+# Generate a common build script for most SuiteSparse packages.
+# use_omp=true will enable OpenMP, this may default to true in the future.
+# use_cuda=true will enable the CUDA CHOLMOD and CUDA SPQR builds,
+# but requires additional set up found in ../SuiteSparse_GPU.
+# CMAKE_OPTIONS prepended to this script can be used to pass additional arguments
+# for instance -DSUITESPARSE_USE_SYSTEM_*=ON to use pre-existing JLLs for
+# certain packages.
+# Use PROJECTS_TO_BUILD to specify which projects to build.
+function build_script(use_omp::Bool = false, use_cuda::Bool = false)
+    return "USEOMP=$(use_omp)\nUSECUDA=$(use_cuda)\n" * raw"""
+cd $WORKSPACE/srcdir/SuiteSparse
+
+# Needs cmake >= 3.24 provided by jll
+apk del cmake
+
+FLAGS+=(INSTALL="${prefix}" INSTALL_LIB="${libdir}" INSTALL_INCLUDE="${prefix}/include")
+
+BLAS_NAME=blastrampoline
+if [[ "${target}" == *-mingw* ]]; then
+    BLAS_LIB=${BLAS_NAME}-5
+else
+    BLAS_LIB=${BLAS_NAME}
+fi
+
+if [[ ${bb_full_target} == *-sanitize+memory* ]]; then
+    # Install msan runtime (for clang)
+    cp -rL ${libdir}/linux/* /opt/x86_64-linux-musl/lib/clang/*/lib/linux/
+fi
+
+if [[ ${nbits} == 64 ]]; then
+    CMAKE_OPTIONS+=(
+        -DBLAS64_SUFFIX="_64"
+        -DSUITESPARSE_USE_64BIT_BLAS=YES
+    )
+else
+    CMAKE_OPTIONS+=(
+        -DSUITESPARSE_USE_64BIT_BLAS=NO
+    )
+fi
+
+# some of these are not used for a particular builder
+# but most are. BLAS handling is the big one which isn't always used.
+# It's likely easier to keep them in common.jl than add them elsewhere.
+# TODO: Can BLAS handling be done by upstream now?
+mkdir -p build && cd build
+cmake -DCMAKE_BUILD_TYPE=Release \
+      -DCMAKE_INSTALL_PREFIX=${prefix} \
+      -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+      -DBUILD_STATIC_LIBS=OFF \
+      -DBUILD_TESTING=OFF \
+      -DSUITESPARSE_ENABLE_PROJECTS=${PROJECTS_TO_BUILD} \
+      -DSUITESPARSE_DEMOS=OFF \
+      -DSUITESPARSE_USE_STRICT=ON \
+      -DSUITESPARSE_USE_FORTRAN=OFF \
+      -DSUITESPARSE_USE_OPENMP=${USEOMP} \
+      -DSUITESPARSE_USE_CUDA=${USECUDA} \
```

Comment:
Looking at this common script, why keep this option here instead of making it an extra flag from `SuiteSparse_GPU`? There is nothing else in this script that uses this `USECUDA` environment variable. And actually, you are already passing this in as an additional variable in the GPU script.

Reply:
Default is ON and was erroring due to missing CUDA for some reason (rather than turning it off when missing CUDA. I suspect due to the strict setting). I should remove it from the GPU script